scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinService.java

```diff
@@ -286,16 +308,86 @@ public void createSimulationForPlan(final PlanId planId)
    * @return
    */
   @Override
-  public Map<ActivityInstance, ActivityInstanceId> updatePlanActivities(final PlanId planId, final Plan plan)
+  public Map<ActivityInstance, ActivityInstanceId> updatePlanActivities(
+      final PlanId planId,
+      final Map<SchedulingActivityInstanceId, ActivityInstanceId> idsFromInitialPlan,
+      final MerlinPlan initialPlan,
+      final Plan plan)
   throws IOException, NoSuchPlanException, MerlinServiceException
   {
-    //TODO: (api violation) currently clearing and repopulating plan; but loses existing activity instance ids!
-    //TODO: (perf improvement) calculate or cache plan diffs during sched and then upload to aerie in batch here
-    //TODO: (defensive) should combine all mutations into one graphql transaction to avoid intermediate mods
-    clearPlanActivities(planId);
-    return createAllPlanActivities(planId, plan);
+    final var ids = new HashMap<ActivityInstance, ActivityInstanceId>();
+    //creation are done in batch as that's what the scheduler does the most
+    final var toAdd = new ArrayList<ActivityInstance>();
+    for (final var activity : plan.getActivities()) {
+      final var idActFromInitialPlan = idsFromInitialPlan.get(activity.getId());
+      if (idActFromInitialPlan != null) {
+        final var actFromInitialPlan = initialPlan.getActivityById(idActFromInitialPlan);
+        //if act was present in initial plan
+        final var schedulerActIntoMerlinAct = new MerlinActivityInstance(activity.getType().getName(), activity.getStartTime(), activity.getArguments());
+        final var activityInstanceId = idsFromInitialPlan.get(activity.getId());
+        if(!schedulerActIntoMerlinAct.equals(actFromInitialPlan.get())) {
+          //update if it has changed
+          updateActivity(planId, schedulerActIntoMerlinAct, activityInstanceId);
+        }
+        ids.put(activity, activityInstanceId);
+      } else {
+        //act was not present in initial plan, create new activity
+        toAdd.add(activity);
+      }
+    }
+    final var actsFromNewPlan = plan.getActivitiesById();
+    for (final var idActInInitialPlan : idsFromInitialPlan.keySet()) {
```

Comment:
IntelliJ is telling me that we can use `entrySet()` instead of `keySet()`, so we won't need to call `.get()` below.